### DISCUSSION TITLE
keep the table grid lines align when use Chinese fonts

### DISCUSSION THREAD

**Discussion by l1t1** [pymupdf/PyMuPDF#3939] [Looking for help] [answered]
**Is your feature request related to a problem? Please describe.**
I want to the output the query result of duckdb to a pdf file.
when I use font name "Courier", It cannot output the grid lines and Chinese charactors, so I visit 
 https://github.com/pymupdf/PyMuPDF/issues/329 and use "china-s",  it shows the output, but the  grid lines are not aligned.

**Describe the solution you'd like**
It output to pdf like windows CMD or NotePad does.

**Describe alternatives you've considered**
Are there several options for how your request could be met?

**Additional context**
```python
import pymupdf
import duckdb
text = duckdb.sql("select '张三' 名字") # some text
doc = pymupdf.open() # empty pdf
text1=str(text)
print(text1)
n = doc.insert_page(-1, # default insertion point
                    text = text1, 
                    fontsize = 11,
                    width = 595,
                    height = 842,
                    fontname = "china-s", # or "Courier"
                    fontfile = None, # any font file name
                    color = (0, 0, 0)) # text color (RGB)
doc.save(
    "duckdboutput.pdf",
)

it output
┌─────────┐
│  名字   │
│ varchar │
├─────────┤
│ 张三    │
└─────────┘
```
![2024-10-12_104409](https://github.com/user-attachments/assets/e3a0fce7-4ebc-48c2-85d5-ee8aaf459b9b)
![2024-10-12_104157](https://github.com/user-attachments/assets/62d5ea4d-e81c-41a7-b34d-aecb307ca39a)
[duckdboutput.pdf](https://github.com/user-attachments/files/17348645/duckdboutput.pdf)

**Comment by JorjMcKie** [comment]
This is not an enhancement request.
You are asking for advice on how to draw table structures and fill them with some text.
Therefore transferring this to a Discussions item.

**Comment by JorjMcKie** [comment]
In PyMuPDF, there are several ways to do this.

* Option 1: Define a table within a given rectangle of a page, draw the cell gridlines and fill them with arbitrary text using `page.insert_htmlbox()` for each table cell.

* Option 2: Define a table completely in HTML syntax and output this on a selected rectangle on the page using `page.insert_htmlbox()`. In this case, table cell dimensioning and all the styling is done automatically.

Example for option 1:
```python
import pymupdf

doc = pymupdf.open()
page = doc.new_page()
rect = pymupdf.Rect(100, 100, 400, 400)
cols = 3
rows = 5
cells = pymupdf.make_table(rect, rows=rows, cols=cols)
for i in range(rows):
    for j in range(cols):
        page.draw_rect(cells[i][j], color=(1, 0, 0))
        page.insert_htmlbox(cells[i][j], f"第 {i} 行、第 {j} 列的任意文本")
doc.ez_save("x.pdf")

```

**Comment by l1t1** [comment]
it's not a draw table question. 
as text1 is a plain text string.

**Reply by JorjMcKie** [reply]
Well, then pass that string into `page.insert_htmlbox()` and see what happens.

**Comment by l1t1** [comment]
@JorjMcKie
thanks for your options, I need to convert  the results to a list or a data frame first to use it.

**Comment by l1t1** [comment]
I tried
```python

page = doc.new_page()
page.insert_htmlbox(page.rect,text1)
```
it output the text without new lines, in other words, it omits the \n chars in the string.

**Reply by JorjMcKie** [reply]
do `text = text.replace("\n","<br>")`.

**Reply by JorjMcKie** [reply]
the method interprets the passed-in text as html syntax.

**Comment by l1t1** [comment]
`page.insert_htmlbox(page.rect,text1.replace("\n","<br>"))` output multiple lines, but it's gridlines are still not aligned. 
btw, the file size increased to 3 MB from original 2KB.

**Reply by JorjMcKie** [reply]
Use `doc.subset_fonts()` before you save using `doc.ez_save()`.
Also recommend to read the documentation!

**Reply by JorjMcKie** [reply]
Lines are no text. So don't be surprised if text strings that contain Unicode characters that **_look like_** line elements do not follow an alignment by some magic.
In your input text, you must differentiate between actual text components and other Unicodes that should be interpreted as graphical.

**Comment by l1t1** [comment]
btw, by using wkhtmltopdf tools can output a fine aligned pdf. --not yet,see
`\wkhtmltox\bin\wkhtmltopdf \htm.txt wkhtmltopdf.pdf` 
[wkhtmltopdf.pdf](https://github.com/user-attachments/files/17352835/wkhtmltopdf.pdf)

the notepad print to pdf does  output a fine aligned pdf, see
[notepadprint.pdf](https://github.com/user-attachments/files/17352836/notepadprint.pdf)

**Comment by l1t1** [comment]
```python
doc.subset_fonts()
doc.ez_save()
```
reduces the file size, but also cannot display the horizon gridlines. such as
```
┌─────────┐ cannot display this line
│ 名字 │
 varchar 
├─────────┤ cannot display this line
│ 张三 │
└────────  cannot display this line
```

**Comment by l1t1** [comment]
I tried many fonts, 
china-s show all, but the english letters with space
cjk show all, but cannot aligned
ubuntm show only non-Chinese content, and aligned 
![china-s](https://github.com/user-attachments/assets/17a9915f-e621-43ff-ad88-03a4378163bc)
![ubuntm](https://github.com/user-attachments/assets/b1d4db6b-0bdc-4a65-b004-6214e9d3ddb1)
![cjk](https://github.com/user-attachments/assets/775bef11-ff0b-4acb-b474-72be37070438)

**Comment by l1t1** [comment, marked as answer]
I solved the question by using a font file 
`page.insert_text((0,0),text1,fontname="milk",fontfile='d:/milky-term-cn-normal.ttf')`

![milk](https://github.com/user-attachments/assets/8789d84e-dd07-4f8f-97ed-3fd6eb814aca)
